metadata/src/from_into/v14.rs

```diff
@@ -178,17 +226,190 @@ fn v14_to_v15(metadata: v14::RuntimeMetadataV14) -> v15::RuntimeMetadataV15 {
+            extra: *extra,
+        })
     }
 }
+
+fn generate_outer_enums(
+    metadata: &mut v14::RuntimeMetadataV14,
+) -> v15::OuterEnums<scale_info::form::PortableForm> {
+    let call_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeCall"
+        })
+        .expect("RuntimeCall exists in V14; qed");
+
+    let event_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeEvent"
+        })
+        .expect("RuntimeEvent exists in V14; qed");
+
+    let call_ty = call_enum.id.into();
+    let event_ty = event_enum.id.into();
+    let error_ty_id = generate_runtime_error_type(metadata);
+
+    v15::OuterEnums {
+        call_enum_ty: call_ty,
+        event_enum_ty: event_ty,
+        error_enum_ty: error_ty_id.into(),
+    }
+}
+
+/// Generate the `RuntimeError` type and add it to the metadata.
+///
+/// Returns the `RuntimeError` Id from the registry.
+fn generate_runtime_error_type(metadata: &mut v14::RuntimeMetadataV14) -> u32 {
+    let error_types: HashMap<_, _> = metadata
+        .types
+        .types
+        .iter()
+        .filter_map(|ty| {
+            let segments = &ty.ty.path.segments;
+            // Interested in segments that end with `pallet::Error`.
+            let len = segments.len();
+            if len < 2 {
+                return None;
+            }
+
+            if segments[len - 2] != "pallet" || segments[len - 1] != "Error" {
+                return None;
+            }
+
+            let pallet_name = segments[0].as_str();
+            let pallet_name = if let Some(name) = pallet_name.strip_prefix("pallet_") {
+                name.to_lowercase()
+            } else if let Some(name) = pallet_name.strip_prefix("frame_") {
+                name.to_lowercase()
+            } else {
+                pallet_name.to_lowercase()
+            };
+
+            Some((pallet_name, ty))
+        })
+        .collect();
+
+    let variants: Vec<_> = metadata
+        .pallets
+        .iter()
+        .filter_map(|pallet| {
+            let Some(entry) = error_types.get(&pallet.name.to_lowercase()) else { return None };
+
+            Some(scale_info::Variant {
+                name: pallet.name.clone(),
+                fields: vec![scale_info::Field {
+                    name: None,
+                    ty: entry.id.into(),
+                    type_name: Some(format!("pallet_{}::Error<Runtime>", pallet.name)),
+                    docs: vec![],
+                }],
+                index: pallet.index,
+                docs: vec![],
+            })
+        })
+        .collect();
+
+    let error_type = scale_info::Type {
+        path: scale_info::Path {
+            segments: vec![
+                "kitchensink_runtime".to_string(),
```

Comment:
why is this injected? 

to work with node-template/node?

Reply:
Subxt uses the latest stable metadata V15 under the hood, just to make sure we expose the latest features (runtime API etc).

Prior to this PR, codegen generated the RuntimeError type manually. For V15 this is already exposed.
We still want to want subxt compatible with V14. When we are targeting a V14 node, we are generating the type here to give subxt a working V15 :D

Reply:
Do we currently test that things work fine if V14 metadata is provided (and needs to be upgraded to V15) rather than just when V15 metadata is provided directly? It'd be good to have confidence that everything works as expected in both cases, ie the shims we insert when upgrading to V15 all work as expected (maybe this is caught in the act of upgrading metadata into the `Metadata` struct?)

Reply:
Ops, missed this thread! Have added some extra tests to check we propagate correctly the newly constructed types (extrinsic / outerenums). While going over the testing, have simplified a bit how we generate the `RuntimeError` variants